sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -736,6 +736,22 @@ abstract class TypeCoercionBase {
     }
   }
 
+  /**
+   * Determines the value type of a [[Unpivot]].
+   */
+  object UnpivotCoercion extends Rule[LogicalPlan] {
```

Comment:
Sorry I missed this. This rule should extend `TypeCoercionRule` like all other rules in this file.

Reply:
Yes, most other rules in that file are `TypeCoercionRule`, except for `WidenSetOperationTypes`.

Like `WidenSetOperationTypes`, `UnpivotCoercion` handles expressions of specific `LogicalPlan`s only (to be precise the `values` expression of `Unpivot` `LogicalPlan` only), not any expression of some kind. So I think it has to hook into `Rule[LogicalPlan].apply` rather than `TypeCoercionRule.transform`.

I have moved `UnpivotCoercion` in that file next to `WidenSetOperationTypes` as they are similar in that sense.